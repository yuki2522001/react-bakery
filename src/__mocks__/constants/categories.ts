import { CategoryProps } from "@common-types/category";

const CATEGORY_MOCKING_LIST: CategoryProps[] = [
  {
    id: "1",
    name: "Cake"
  },
  {
    id: "2",
    name: "Breads"
  }
]

export { CATEGORY_MOCKING_LIST };
