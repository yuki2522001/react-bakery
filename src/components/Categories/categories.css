.categories {
  margin-top: 50px;
  text-align: center;
}

.categories__title {
  text-transform: uppercase;
  font-weight: bold;
}

.categories__list {
  display: flex;
  justify-content: center;
  margin: 30px 0 40px 0;
  padding: 0;
}

.categories__item {
  list-style: none;
  cursor: pointer;
  margin: 0 30px;
  position: relative;
  color: var(--text-secondary);
}

.categories__item:hover {
  color: var(--text-primary);
}

.active {
  color: var(--text-primary);
}

.categories__item > a {
  color: var(--text-third);
  text-decoration: none;
}