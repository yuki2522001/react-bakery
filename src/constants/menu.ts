import { MenuType } from "../common-types/menu";

const MENU_LIST: MenuType[] = [
  {
    key: "about",
    label: "About",
  },
  {
    key: "shop",
    label: "Shop",
  },
  {
    key: "blog",
    label: "Blog",
  },
  {
    key: "faq",
    label: "FAQ",
  },
  {
    key: "contact",
    label: "Contact"
  }
];

export default MENU_LIST;
