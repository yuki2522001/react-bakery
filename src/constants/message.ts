const SUCCESS_MSG = {
  MESSAGE_ADD_PRODUCT: "Product add successful",
  MESSAGE_UPDATE_PRODUCT: "Product update successful",
  MESSAGE_DELETE_PRODUCT: "Product delete successful",
};

const ERROR_MSG = {
  REQUIRED: "Field is required",
  NUMBER: "Enter a non-negative number"
}

export { SUCCESS_MSG, ERROR_MSG };
