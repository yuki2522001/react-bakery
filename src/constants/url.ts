const PRODUCTS_URL = `${process.env.VITE_BASE_URL}/products`;
const CATEGORIES_URL = `${process.env.VITE_BASE_URL}/categories`;

export { PRODUCTS_URL, CATEGORIES_URL };