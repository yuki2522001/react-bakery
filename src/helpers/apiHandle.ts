import axios, { AxiosResponse } from "axios";

// Get data form server
const getData = async <T>(url: string): Promise<T> => {
  const res = await axios.get<T>(url);
  return res.data;
};

// Post new data to server
const create = async <T>(url: string, newData: T):Promise<AxiosResponse<T>> => {
  const res = await axios.post(url, newData);
  return res;
}

// Send a changed data to server using method put to updating the data
const update = async <T>(id: string, dataEdit: T):Promise<AxiosResponse<T>> => {
  const res = await axios.put(id, {...dataEdit});
  return res;
}

// Delete data
const remove = async <T>(id: string): Promise<AxiosResponse<T>> => {
  const res = await axios.delete(id);
  return res;
}

export {
  getData,
  create,
  update,
  remove
}