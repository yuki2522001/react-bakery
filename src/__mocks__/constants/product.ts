import { Product } from "@common-types/product";
import url from "../../assets/images/products/black-forest-cake.jpg";

const PRODUCT_MOCKING: Product = {
  id: "1",
  categoryId: "1651924356334",
  name: "Cheese cake",
  price: 20,
  quantity: 5,
  description: "Sed ut perspiciatis unde omnis iste natus error sit voluptatem accusantium doloremque laudantium.",
  images: [
    url
  ]
}

const PRODUCT_MOCKING_LIST: Product[] = [
  {
    id: "1",
    categoryId: "1651924356334",
    name: "Cheese cake",
    price: 20,
    quantity: 5,
    description: "Sed ut perspiciatis unde omnis iste natus error sit voluptatem accusantium doloremque laudantium.",
    images: [
      "/src/assets/images/products/cheese-cake.jpg",
      "/src/assets/images/products/cheese-cake.jpg",
      "/src/assets/images/products/cheese-cake.jpg"
    ]
  },
  {
    id: "2",
    categoryId: "1651924356334",
    name: "Amaretti Cookie",
    price: 10,
    quantity: 5,
    description: "Sed ut perspiciatis unde omnis iste natus error sit voluptatem accusantium doloremque laudantium.",
    images: [
      "/src/assets/images/products/cheese-cake.jpg",
      "/src/assets/images/products/cheese-cake.jpg",
      "/src/assets/images/products/cheese-cake.jpg"
    ]
  }
]

export { PRODUCT_MOCKING, PRODUCT_MOCKING_LIST };
