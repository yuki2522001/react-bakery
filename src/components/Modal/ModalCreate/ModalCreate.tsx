import useSWR from "swr";
import React, { ChangeEvent, useCallback, useState } from "react";
import getBase64 from "@helpers/getBase64";
import InputValue from "@components/Input/InputValue/InputValue";
import Button from "@components/common/Button/Button/Button";
import { FieldName, FormProps } from "@common-types/form";
import { Product } from "@common-types/product";
import { CategoryProps } from "@common-types/category";
import { RULES } from "@constants/rules";
import { CATEGORIES_URL } from "@constants/url";
import { getData } from "@helpers/apiHandle";
import { validate } from "@helpers/validate";
import { setFieldsValue } from "@helpers/fieldHandle";
import "../modal.css";

interface ModalCreateProps {
  hideModalCreate: () => void;
  createProduct: (dataProduct: Product) => void;
}

const ModalCreate: React.FC<ModalCreateProps> = ({
  hideModalCreate,
  createProduct,
}) => {
  const initProduct: Product = {
    id: "",
    name: "",
    price: 0,
    images: [],
    quantity: 0,
    categoryId: "",
    description: ""
  }
  const [isDisable, setIsDisable] = useState<boolean>(false);
  const [newProduct, setNewProduct] = useState<Product>(initProduct);
  // fetch data with useSWR
  const { data } = useSWR(CATEGORIES_URL, getData<Product[]>);
  // create state to handle select file image
  const [selectedFile, setSelectedFile] = useState<string[]>([]);
  // create state to set form values
  const [formValues, setFormValues] = useState<FormProps>({
    categoryId: {
      value: "",
      rules: [RULES.REQUIRED],
      error: "",
    },
    name: {
      value: "",
      rules: [RULES.REQUIRED],
      error: "",
    },
    price: {
      value: "",
      rules: [RULES.REQUIRED, RULES.NUMBER, RULES.NEGATIVE],
      error: "",
    },
    quantity: {
      value: "",
      rules: [RULES.REQUIRED, RULES.NUMBER, RULES.NEGATIVE],
      error: "",
    },
    description: {
      value: "",
      rules: [RULES.REQUIRED],
      error: "",
    },
    images: {
      value: "",
      rules: [],
      error: "",
    }
  });

  // handle create product
  const handleCreateProduct = () => {
    const images: string[] = [];
    for (let i = 0; i < selectedFile.length; i++) {
      images.push(selectedFile[i]);
    }
    // validate form
    setFormValues(validate(formValues));
    const temp = (Object.keys(formValues) as (keyof typeof formValues)[]).map(
      (fieldName) => {
        if (formValues[fieldName].error) {
          return false;
        }
      }
    );

    // check validate if pass then create product
    if (!temp.includes(false)) {
      setFormValues({...formValues});
      setIsDisable(!isDisable);
      const product: Product = { ...newProduct, images: images }
      createProduct(product);
      hideModalCreate();
    }
  };

  // handle change value
  const handleChange = useCallback((event: { target: { value: string; name: string } }) => {
    const value = event.target.value;
    const fieldName = event.target.name as FieldName;
    setNewProduct({ ...newProduct, [fieldName]: value });
    setFormValues(setFieldsValue(formValues, value, fieldName));
  }, [formValues]);

  // handle change image
  const imageChange = useCallback(async (event: ChangeEvent<HTMLInputElement>) => {
    const files = event.target.files;
    const newFiles = [];
    if (files) {
      for (let i = 0; i < files.length; i++) {
        const imageSrc = await getBase64(files[i]);
        newFiles.push(imageSrc)
      }
    }
    setSelectedFile([...selectedFile, ...newFiles]);
  }, [selectedFile]);

  // handle delete image
  const handleDeleteImage = useCallback((event: { target: EventTarget }) => {
    const target = event.target as Element;
    const indexOfArr = selectedFile.findIndex(
      (item: string) => item == (target as HTMLInputElement).dataset.id
    );
    selectedFile.splice(indexOfArr, 1);
    setSelectedFile([...selectedFile]);
  }, [selectedFile]);

  return (
    <div
      data-testid="modal-create"
      className="modal-create"
      id="bookDeleteModal"
    >
      <div className="modal-dialog-modalUpdate">
        <div className="modal-content">
          <div className="modal-header">
            <h4 className="modal-heading" id="productModalLabel">
              Add New Product Information
            </h4>
          </div>
          <div className="modal-body">
            <div className="form-control">
              <label htmlFor="">Product name <span className="star">*</span></label>
              <InputValue
                type="text"
                name="name"
                onChange={handleChange}
              />
              <small className="form__error">
                {formValues?.name?.error ? formValues.name.error : ""}
              </small>
            </div>
            <div className="form-control">
              <label htmlFor="">Description <span className="star">*</span></label>
              <textarea
                data-testid="change-value"
                className="input__text"
                name="description"
                id=""
                cols={30}
                rows={5}
                onChange={handleChange}
              ></textarea>
              <small className="form__error">
                {formValues?.description?.error
                  ? formValues.description.error
                  : ""}
              </small>
            </div>
            <div className="form-control">
              <label htmlFor="">Categories <span className="star">*</span></label>
              <select
                className="form__select"
                name="categoryId"
                defaultValue={""}
                onChange={handleChange}
              >
                <option value="" disabled>Choose a category ...</option>
                {data?.map(({ id, name }: CategoryProps, index: number) => (
                  <option key={index} value={id}>
                    {name}
                  </option>
                ))}
              </select>
              <small className="form__error">
                {formValues?.categoryId?.error
                  ? formValues.categoryId.error
                  : ""}
              </small>
            </div>
            <div id="form__number" className="form-control">
              <div className="form-control">
                <label htmlFor="">Price <span className="star">*</span></label>
                <InputValue
                  className="input__number"
                  type="number"
                  min={0}
                  name="price"
                  onChange={handleChange}
                />
                <small className="form__error">
                  {formValues?.price?.error ? formValues.price.error : ""}
                </small>
              </div>
              <div className="form-control">
                <label htmlFor="">Quantity <span className="star">*</span></label>
                <InputValue
                  className="input__number"
                  type="number"
                  min={0}
                  name="quantity"
                  onChange={handleChange}
                />
                <small className="form__error">
                  {formValues?.quantity?.error ? formValues.quantity.error : ""}
                </small>
              </div>
            </div>
            <div className="form-control">
              <div className="form__img--list">
                {selectedFile.length > 0 &&
                  selectedFile.map((src, key: number) => (
                    <img
                      data-testid="after-change-file"
                      className="form__img"
                      key={key}
                      data-id={src}
                      src={src}
                      onClick={handleDeleteImage}
                    />
                  ))}
                <input
                  data-testid="change-file"
                  className="form__input--img"
                  type="file"
                  id="file"
                  multiple
                  name="images"
                  onChange={imageChange}
                />
              </div>
            </div>
          </div>
          <div className="modal-footer-modalUpdate">
            <Button
              type="warning"
              text="Cancel"
              onClick={hideModalCreate}
            />
            <Button
              type="success"
              text="Submit"
              disabled = {isDisable}
              onClick={() => handleCreateProduct()}
            />
          </div>
        </div>
      </div>
    </div>
  );
};

export default ModalCreate;
