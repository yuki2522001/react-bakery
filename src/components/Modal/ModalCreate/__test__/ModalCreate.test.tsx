import "@testing-library/jest-dom";
import mockAxios from "@__mocks__/axios";
import user from "@testing-library/user-event";
import ModalCreate from "../ModalCreate";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import { useState } from "react";
import { CATEGORY_MOCKING_LIST } from "@__mocks__/constants/categories";
import { CATEGORIES_URL } from "@constants/url";
import { getData } from "@helpers/apiHandle";

jest.mock("react", () => ({
  ...jest.requireActual("react"),
  useState: jest.fn(),
}));

describe("Modal create component", () => {
  const hideModalCreate = jest.fn();
  const handleCreateProduct = jest.fn();
  const someValues = [{ name: "cheese pocket" }];
  const setup = () => {
    const utils = render(
      <ModalCreate
        hideModalCreate={() => {}}
        createProduct={() => {}}
      />
    );
    const input = utils.getByTestId("change-value") as HTMLInputElement;
    return {
      input,
      ...utils,
    };
  };

  beforeEach(() => {
    (useState as jest.Mock).mockImplementation(
      jest.requireActual("react").useState
    );
  });

  afterEach(() => {
    mockAxios.reset();
  });

  test("should change value when onChang input", () => {
    const { input } = setup();
    fireEvent.change(input, { target: { value: "Cheese pocket" } });
    expect(input.value).toBe("Cheese pocket");
  });

  test("should change file image when onChang input", async () => {
    const { getByTestId, queryByTestId } = render(
      <ModalCreate
        hideModalCreate={() => {}}
        createProduct={() => {}}
      />
    );
    const str = JSON.stringify(someValues);
    const blob = new Blob([str]);
    const file = new File([blob], "db.json", {
      type: "application/JSON",
    });
    File.prototype.text = jest.fn().mockResolvedValueOnce(str);
    const input = getByTestId("change-file");
    user.upload(input, file);
    await waitFor(() =>
      expect(queryByTestId("after-change-file")).toBeTruthy()
    );
  });

  test("get categories item should call", async () => {
    mockAxios.get.mockResolvedValueOnce({ data: CATEGORY_MOCKING_LIST });
    const result = await getData(CATEGORIES_URL);
    expect(mockAxios.get).toHaveBeenCalledWith(CATEGORIES_URL);
    expect(result).toEqual(CATEGORY_MOCKING_LIST);
  });

  test("should render modal create component", () => {
    const { getByTestId } = render(
      <ModalCreate
        hideModalCreate={() => {}}
        createProduct={() => {}}
      />
    );
    expect(getByTestId("modal-create")).toBeInTheDocument();
  });

  test("should hide modal create when click Cancel", () => {
    render(
      <ModalCreate
        hideModalCreate={hideModalCreate}
        createProduct={() => {}}
      />
    );
    const cancelBtn = screen.getByText("Cancel");
    fireEvent.click(cancelBtn);
    expect(hideModalCreate).toHaveBeenCalled();
  });

  test("should create product when click Submit", () => {
    render(
      <ModalCreate
        hideModalCreate={() => {}}
        createProduct={handleCreateProduct}
      />
    );
    const submitBtn = screen.getByText("Submit");
    fireEvent.click(submitBtn);
    expect(submitBtn).toBeInTheDocument();
  });

  test("matches snapshot", () => {
    const { asFragment } = render(
      <ModalCreate
        hideModalCreate={() => {}}
        createProduct={() => {}}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
