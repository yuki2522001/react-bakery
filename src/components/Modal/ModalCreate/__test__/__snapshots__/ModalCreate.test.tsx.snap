// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal create component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="modal-create"
    data-testid="modal-create"
    id="bookDeleteModal"
  >
    <div
      class="modal-dialog-modalUpdate"
    >
      <div
        class="modal-content"
      >
        <div
          class="modal-header"
        >
          <h4
            class="modal-heading"
            id="productModalLabel"
          >
            Add New Product Information
          </h4>
        </div>
        <div
          class="modal-body"
        >
          <div
            class="form-control"
          >
            <label
              for=""
            >
              Product name 
              <span
                class="star"
              >
                *
              </span>
            </label>
            <input
              name="name"
              type="text"
              value=""
            />
            <small
              class="form__error"
            />
          </div>
          <div
            class="form-control"
          >
            <label
              for=""
            >
              Description 
              <span
                class="star"
              >
                *
              </span>
            </label>
            <textarea
              class="input__text"
              cols="30"
              data-testid="change-value"
              id=""
              name="description"
              rows="5"
            />
            <small
              class="form__error"
            />
          </div>
          <div
            class="form-control"
          >
            <label
              for=""
            >
              Categories 
              <span
                class="star"
              >
                *
              </span>
            </label>
            <select
              class="form__select"
              name="categoryId"
            >
              <option
                disabled=""
                selected=""
                value=""
              >
                Choose a category ...
              </option>
            </select>
            <small
              class="form__error"
            />
          </div>
          <div
            class="form-control"
            id="form__number"
          >
            <div
              class="form-control"
            >
              <label
                for=""
              >
                Price 
                <span
                  class="star"
                >
                  *
                </span>
              </label>
              <input
                class="input__number"
                min="0"
                name="price"
                type="number"
                value=""
              />
              <small
                class="form__error"
              />
            </div>
            <div
              class="form-control"
            >
              <label
                for=""
              >
                Quantity 
                <span
                  class="star"
                >
                  *
                </span>
              </label>
              <input
                class="input__number"
                min="0"
                name="quantity"
                type="number"
                value=""
              />
              <small
                class="form__error"
              />
            </div>
          </div>
          <div
            class="form-control"
          >
            <div
              class="form__img--list"
            >
              <input
                class="form__input--img"
                data-testid="change-file"
                id="file"
                multiple=""
                name="images"
                type="file"
              />
            </div>
          </div>
        </div>
        <div
          class="modal-footer-modalUpdate"
        >
          <button
            class="btn btn__warning"
          >
            Cancel
          </button>
          <button
            class="btn btn__success"
          >
            Submit
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
